pkg/migrations/sql/idempotent/tracing.sql

```diff
@@ -0,0 +1,649 @@
+RETURNS TABLE
+(
+    trace_id _ps_trace.trace_id,
+    parent_span_id bigint,
+    span_id bigint,
+    dist int,
+    idx int,
+    path bigint[]
+)
+AS $func$
+    SELECT
+        trace_id,
+        parent_span_id,
+        span_id,
+        dist * -1 as dist,
+        idx,
+        path
+    FROM _ps_trace.upstream_spans(_trace_id, _span_id, _max_dist)
+    UNION
+    SELECT
+        trace_id,
+        parent_span_id,
+        span_id,
+        dist as dist,
+        idx,
+        path
+    FROM _ps_trace.downstream_spans(_trace_id, _span_id, _max_dist)
+$func$ LANGUAGE sql STABLE;
+GRANT EXECUTE ON FUNCTION _ps_trace.span_tree(_ps_trace.trace_id, bigint, int) TO prom_reader;
+
+CREATE OR REPLACE FUNCTION _ps_trace.put_tag_key(_key _ps_trace.tag_k, _tag_type _ps_trace.tag_type)
+RETURNS VOID
+AS $func$
+    INSERT INTO _ps_trace.tag_key AS k (key, tag_type)
+    VALUES (_key, _tag_type)
+    ON CONFLICT (key) DO
+    UPDATE SET tag_type = k.tag_type | EXCLUDED.tag_type
+    WHERE k.tag_type & EXCLUDED.tag_type = 0
+$func$
+LANGUAGE SQL VOLATILE STRICT;
+GRANT EXECUTE ON FUNCTION _ps_trace.put_tag_key(_ps_trace.tag_k, _ps_trace.tag_type) TO prom_writer;
+
+CREATE OR REPLACE FUNCTION _ps_trace.put_tag(_key _ps_trace.tag_k, _value _ps_trace.tag_v, _tag_type _ps_trace.tag_type)
+RETURNS VOID
+AS $func$
+    INSERT INTO _ps_trace.tag AS a (tag_type, key_id, key, value)
+    SELECT _tag_type, ak.id, _key, _value
+    FROM _ps_trace.tag_key ak
+    WHERE ak.key = _key
+    ON CONFLICT (key, value) DO
+    UPDATE SET tag_type = a.tag_type | EXCLUDED.tag_type
+    WHERE a.tag_type & EXCLUDED.tag_type = 0
+$func$
+LANGUAGE SQL VOLATILE STRICT;
+GRANT EXECUTE ON FUNCTION _ps_trace.put_tag(_ps_trace.tag_k, _ps_trace.tag_v, _ps_trace.tag_type) TO prom_writer;
+
+CREATE OR REPLACE FUNCTION _ps_trace.has_tag(_tag_map _ps_trace.tag_map, _key _ps_trace.tag_k)
+RETURNS boolean
+AS $func$
+    SELECT _tag_map ?
+    (
+        SELECT k.id::text
+        FROM _ps_trace.tag_key k
+        WHERE k.key = _key
+        LIMIT 1
+    )
+$func$
+LANGUAGE SQL STABLE PARALLEL SAFE STRICT;
+GRANT EXECUTE ON FUNCTION _ps_trace.has_tag(_ps_trace.tag_map, _ps_trace.tag_k) TO prom_reader;
+
+CREATE OR REPLACE FUNCTION _ps_trace.jsonb(_tag_map _ps_trace.tag_map)
+RETURNS jsonb
+AS $func$
+    /*
+    takes an tag_map which is a map of tag_key.id to tag.id
+    and returns a jsonb object containing the key value pairs of tags
+    */
+    SELECT jsonb_object_agg(a.key, a.value)
+    FROM jsonb_each(_tag_map) x -- key is tag_key.id, value is tag.id
+    INNER JOIN LATERAL -- inner join lateral enables partition elimination at execution time
+    (
+        SELECT
+            a.key,
+            a.value
+        FROM _ps_trace.tag a
+        WHERE a.id = x.value::text::bigint
+        AND a.key = (SELECT k.key from _ps_trace.tag_key k WHERE k.id = x.key::bigint)
```

Comment:
It's there for performance reasons. It's faster to leave this clause in, because it enables partition elimination. It will only scan the index on one partition this way, rather than scanning the indexes on every partition of the tag table

Reply:
i'll add a comment to explain